feat: Allow creation of new window to be customisable in order to support window.opener in BrowserView.

khalwa opened the pull request:
#### Description of Change
This PR adds a way to manually control the creation of a guest window in `webContents.setWindowOpenHandler`. My main motivation for this change is to have the possibility to properly handle new windows in form of BrowserView (in order to support `window.opener` I need to pass appropriate `WebContents` when creating an instance of it).

I am not an expert so I don't know what important things I could've missed introducing this change but I am hoping that with your guidance we'll be able to work on this PR and make such functionality possible in Electron. 

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation is changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes
Notes: Extended `webContents.setWindowOpenHandler` to support manual creation of BrowserWindow.

nornagon commented on docs/api/structures/window-open-handler-response.md:
```diff
@@ -0,0 +1,20 @@
+* `action` string - Can be `allow` or `deny`. Controls whether new window should be created.
+* `overrideBrowserWindowOptions` BrowserWindowConstructorOptions (optional) - Allows customization of the created window.
+* `outlivesOpener` boolean (optional) - Controls whether created window should stay open when its opener was closed.
+* `overrideWindowCreationFunction` (options: BrowserWindowConstructorOptions, wc?: WebContents) => BrowserWindow (optional) - Allows to manually control creation of window.
```
```suggestion
* `createWindow` (options: BrowserWindowConstructorOptions) => BrowserWindow (optional) - If specified, will be called instead of `new BrowserWindow` to create the new child window. This can be used for example to have the new window open as a BrowserView instead of in a separate window.
```

Let's rename this and simplify it a little—we can pass in a pure `options` structure, and include the `webContents` parameter in it.

nornagon commented on docs/api/structures/window-open-handler-response.md:
```diff
@@ -0,0 +1,20 @@
+
+* `action` string - Can be `allow` or `deny`. Controls whether new window should be created.
+* `overrideBrowserWindowOptions` BrowserWindowConstructorOptions (optional) - Allows customization of the created window.
+* `outlivesOpener` boolean (optional) - Controls whether created window should stay open when its opener was closed.
```
Thanks for properly documenting this!

nornagon commented on docs/api/structures/window-open-handler-response.md:
````diff
@@ -0,0 +1,20 @@
+* `outlivesOpener` boolean (optional) - Controls whether created window should stay open when its opener was closed.
+* `overrideWindowCreationFunction` (options: BrowserWindowConstructorOptions, wc?: WebContents) => BrowserWindow (optional) - Allows to manually control creation of window.
+
+```javascript
````
Please add some explanation about what this example shows.

nornagon commented on docs/api/web-contents.md:
```diff
@@ -1187,8 +1187,8 @@ Ignore application menu shortcuts while this web contents is focused.
-  Returns `{action: 'deny'} | {action: 'allow', overrideBrowserWindowOptions?: BrowserWindowConstructorOptions}` - `deny` cancels the creation of the new
-  window. `allow` will allow the new window to be created. Specifying `overrideBrowserWindowOptions` allows customization of the created window.
+  Returns `WindowOpenHandlerResponse` -  When set to `{ action: 'deny' }` cancels the creation of the new
+  window. `{ action: 'allow' }` will allow the new window to be created. Specifying `overrideBrowserWindowOptions` allows customization of the created window. Specyfing `overrideWindowCreationFunction` allows to customize the way that window is created.
```
```suggestion
  window. `{ action: 'allow' }` will allow the new window to be created. Specifying `overrideBrowserWindowOptions` allows customization of the created window. Specifying `createWindow` allows to customize the way that window is created.
```

nornagon commented on lib/browser/api/web-contents.ts:
```diff
@@ -522,17 +523,17 @@ WebContents.prototype._callWindowOpenHandler = function (event: Electron.Event,
     event.preventDefault();
     return defaultResponse;
   } else if (response.action === 'allow') {
+    defaultResponse.outlivesOpener = typeof response.outlivesOpener === 'boolean' ? response.outlivesOpener : false;
```
It seems a little odd to be modifying `defaultResponse` here. Can we perhaps name that to `finalResponse`, or otherwise refactor this to avoid modifying `defaultResponse`?

nornagon commented on lib/browser/guest-window-manager.ts:
```diff
@@ -73,10 +74,15 @@ export function openGuestWindow ({ event, embedder, guest, referrer, disposition
-  });
+  let window: BrowserWindow;
+  if (overrideWindowCreationFunction) {
+    window = overrideWindowCreationFunction(browserWindowOptions, guest);
```
```suggestion
    window = overrideWindowCreationFunction({
      webContents: guest,
      ...browserWindowOptions
    });
```

khalwa replied:
Should we leave `webContents: guest` line above the `...browserWindowOptions`? `webContents` value will get overwritten by the value that comes in the `browserWindowOptions` object if `webContents` property of it is set.

nornagon commented on spec-main/guest-window-manager-spec.ts:
```diff
@@ -229,6 +229,72 @@ describe('webContents.setWindowOpenHandler', () => {
+
+    browserWindow.webContents.on('did-create-window', (window) => {
+      expect(window.title).to.equal(browserWindowTitle);
+      done();
```
Please use the `async` test style; tests that run expectations in callbacks cause the whole test suite to exit if this test fails.

nornagon commented on typings/internal-electron.d.ts:
```diff
@@ -101,6 +101,15 @@ declare namespace Electron {
+  
+  type OverrideWindowCreationFunction = (options: BrowserWindowConstructorOptions, wc?: WebContents) => BrowserWindow;
+
+  interface WindowOpenHandlerResponse {
+    action: 'allow' | 'deny';
+    overrideBrowserWindowOptions?: BrowserWindowConstructorOptions;
+    outlivesOpener?: boolean;
+    overrideWindowCreationFunction?: OverrideWindowCreationFunction;
+  }
```
I don't think this needs to be duplicated from `electron.d.ts`?

nornagon reviewed (changes requested):
Thanks for your patience waiting for review on this, I know it's been waiting for a while!

I think this is a reasonable thing to want to do—to have new windows open in BrowserViews rather than in a new BrowserWindow. A few notes on the specifics though!

nornagon commented:
Adding the `wip` label to drop this out of the API WG review queue, I'll unmark it once there's movement here.

arestov commented:
@nornagon how can anyone help to make this api improvement happen?

(https://github.com/electron/electron/issues/33383 looks related)

nornagon commented:
@arestov it seems like the original author of this PR has vanished; feel free to create a new PR to continue where they left off and address the comments that I made here!

khalwa commented:
I'm here, I just did not have the time lately to set up new development environment for Electron to introduce the PR recommendations. I'll try to take care of it in the upcoming days and address the comments.

khalwa commented:
@nornagon I have addressed your PR comments, please review when you'll have time. Also I wonder if I should add `webContents` property to `BrowserWindowConstructorOptions` in `electron.d.ts` typings as this feature kind of relies on setting this property which is implicit currently. The same as in example I put in the documentation, creation of `BrowserView` works because it internally accepts `webContents` property but is listed in typings for `BrowserViewConstructorOptions`.

nornagon commented on docs/api/structures/window-open-handler-response.md:
```diff
@@ -0,0 +1,22 @@
+* `action` string - Can be `allow` or `deny`. Controls whether new window should be created.
+* `overrideBrowserWindowOptions` BrowserWindowConstructorOptions (optional) - Allows customization of the created window.
+* `outlivesOpener` boolean (optional) - Controls whether created window should stay open when its opener was closed.
+* `createWindow` (options: BrowserWindowConstructorOptions) => BrowserWindow (optional) - If specified, will be called instead of `new BrowserWindow` to create the new child window. This can be used for example to have the new window open as a BrowserView instead of in a separate window.
```
I think this should probably return a `WebContents` instead of a `BrowserWindow`. For example, it should be possible to open a child window as a BrowserView inside an existing BrowserWindow. I don't think the calling code requires `BrowserWindow`, since all it does is call `loadURL` I think.

khalwa replied:
`openGuestWindow` which uses this function expects to return a `BrowserWindow | undefined` but nothing is using this return value so I guess returning `undefined` in such case is okay?

nornagon replied:
@SEUH WebContents is the right thing to return here, as that's what the calling function needs. I could see it being renamed `createWebContents` but that isn't quite the right name either, because in some cases it will not create a new WebContents, for example when the caller provides a `webContents` property in the BrowserWindowConstructorOptions, it should be used to create the new window.

khalwa commented:
@nornagon Please take a look at the updates.

nornagon commented on docs/api/structures/window-open-handler-response.md:
````diff
@@ -0,0 +1,22 @@
+* `outlivesOpener` boolean (optional) - Controls whether created window should stay open when its opener was closed.
+* `createWindow` (options: BrowserWindowConstructorOptions) => WebContents (optional) - If specified, will be called instead of `new BrowserWindow` to create the new child window. This can be used for example to have the new window open as a BrowserView instead of in a separate window.
+
+An example showing how to customize the process new `BrowserWindow` creation to be `BrowserView` attached to main window instead:
+
+```javascript
+webContents.setWindowOpenHandler((details) => {
+  return {
+    action: 'allow',
+    createWindow: (options) => {
+      const browserView = new BrowserView(options)
+      mainWindow.addBrowserView(browserView)
+      browserView.setBounds({ x: 0, y: 0, width: 640, height: 480 })
+      return browserView.webContents
+    }
+  }
+})
+```
````
Nice example! Mind moving it to the `setWindowOpenHandler` documentation?

nornagon commented on lib/browser/guest-window-manager.ts:
```diff
@@ -59,6 +60,23 @@ export function openGuestWindow ({ embedder, guest, referrer, disposition, postD
+      })
+    });
+
+    return undefined;
```
It looks like nothing ever uses the result of `openGuestWindow`, so let's change the return type to `void` and make this just
```suggestion
    return;
```

nornagon commented on lib/browser/guest-window-manager.ts:
```diff
@@ -59,6 +60,23 @@ export function openGuestWindow ({ embedder, guest, referrer, disposition, postD
+      ...browserWindowOptions
+    });
+
+    webContents.loadURL(url, {
+      httpReferrer: referrer,
+      ...(postData && {
+        postData,
+        extraHeaders: formatPostDataHeaders(postData as Electron.UploadRawData[])
+      })
+    });
```
the code below only does this `if (!guest)`, we should do that here too.

I think we also want to do `handleWindowLifecycleEvents`, as that's what makes `outlivesOpener` work correctly. That method will have to be refactored to take a WebContents instead of a BrowserWindow, but I think that is better actually: use the `destroyed` event of the `WebContents` instead of the `closed` event of `BrowserWindow`.

We can't emit `did-create-window` since that event takes a BrowserWindow as its first argument, but I think that's okay, since it would happen at the same time that the `createWindow` function is called anyway, so it wouldn't provide any additional capability. We should probably document that that event won't be emitted if you specify `createWindow` though.

nornagon commented on spec/guest-window-manager-spec.ts:
```diff
@@ -2,195 +2,330 @@ import { BrowserWindow } from 'electron';
+
+      await once(childWindow.webContents, 'ready-to-show');
+      const childWindowOpenerTitle = await childWindow.webContents.executeJavaScript('window.opener.document.title');
+      expect(browserWindow.title).to.equal(childWindowOpenerTitle);
```
value under test goes first, expected value goes last
```suggestion
      expect(childWindowOpenerTitle).to.equal(browserWindow.title);
```

nornagon commented on spec/guest-window-manager-spec.ts:
```diff
@@ -2,195 +2,330 @@ import { BrowserWindow } from 'electron';
+      });
+
+      await once(childWindow.webContents, 'ready-to-show');
+      const accessingOpenerThrownException = await new Promise((resolve) => {
+        childWindow.webContents.executeJavaScript('window.opener.document.title')
+          .then(() => resolve(false))
+          .catch(() => resolve(true));
+      });
+      expect(accessingOpenerThrownException).to.equal(true);
```
```suggestion
      await expect(childWindow.webContents.executeJavaScript('window.opener.document.title')).to.eventually.be.rejectedWith("<error message>")
```

nornagon reviewed (commented):
Thanks for working on this :) a few more comments below, mainly around the lifecycle / outlivesOpener handling.

nornagon commented on docs/api/web-contents.md:
```diff
@@ -1225,11 +1225,11 @@ Ignore application menu shortcuts while this web contents is focused.
-  changed by specifying `outlivesOpener: true`, in which case the opened window
-  will not be closed when its opener is closed.
+  changed by specifying `outlivesOpener: true`, in which case the opened window will not be closed when its opener is closed.
+  Specyfing `createWindow` allows to customize the way that window is created.
```
```suggestion
  Specifying `createWindow` allows to customize the way that window is created.
```

nornagon commented on lib/browser/guest-window-manager.ts:
```diff
@@ -59,6 +60,23 @@ export function openGuestWindow ({ embedder, guest, referrer, disposition, postD
+    const webContents = createWindow({
+      webContents: guest,
+      ...browserWindowOptions
+    });
```
Do we want to maybe throw an error here if `guest` is non-null and `webContents !== guest`?

i.e. a WebContents was passed in, but `createWindow` chose to return a different WebContents. I think that might cause a memory leak or other undesirable behavior, as the `guest` WebContents would still be hanging around. Especially, the browser may still navigate the `guest` WebContents even if it isn't installed into a window from `createWindow`.

e.g.
```js
const w = new BrowserWindow
w.webContents.setWindowOpenHandler((details) => ({
  action: 'allow',
  createWindow(opts) {
    const childWindow = new BrowserWindow() // no opts
    return childWindow.webContents // this is a fresh WebContents
  }
}))
await w.loadURL(`data:text/html,<script>
  const child = window.open('about:blank')
  child.document.documentElement.innerHTML = '<h1>hello</h1>'
</script>`)
```

In the above scenario, I think the behavior of this PR currently is that a blank child window would be opened, but because `guest` wasn't used when creating the child window, it would not be properly linked to the parent window. However, the child WebContents would still exist, it just wouldn't be displayed anywhere.

I think that's pretty surprising behavior and we should probably help developers avoid that situation. I'm not sure there's a really good argument to be made that it should be legal to return a WebContents that's different to the `{ webContents }` that's passed to `createWindow`, if there is one—if you want to cancel the window.open and create an unconnected WebContents, you can return `action: 'deny'` and create the unconnected child window yourself.

PascalPixel commented:
This is awesome! Come on @khalwa home stretch! 🔥

douglchambers commented:
Love this PR, it would make some things I'm working on much, much easier, is there any movement on it?

khalwa commented:
I'll try to get back to this PR this week (or weekend) as I have quite busy schedule lately.

khalwa commented:
@nornagon I've introduced requested changes. Please take a look and let me know if some further changes are needed.

arestov commented:
@khalwa I would like to express my greatest appreciation that you keep pushing. Thank you so much! ❤️❤️❤️

douglchambers commented:
@nornagon - is it possible to get this one back into the review queue? It would be incredibly helpful for some things I'm trying to do. Thanks.

khalwa commented:
@nornagon Gentle reminder about this PR :smile:.

codebytere commented:
Hi all - `@nornagon` is on extended leave right now. He will be back in a month or so - please do not keep @ mentioning him!

khalwa commented:
@nornagon Do you have the time to continue the code review on this PR?

nornagon commented on docs/api/web-contents.md:
```diff
@@ -1302,11 +1302,11 @@ Ignore application menu shortcuts while this web contents is focused.
-  window. `allow` will allow the new window to be created. Specifying `overrideBrowserWindowOptions` allows customization of the created window.
+  Returns `WindowOpenHandlerResponse` -  When set to `{ action: 'deny' }` cancels the creation of the new
+  window. `{ action: 'allow' }` will allow the new window to be created. Specifying `overrideBrowserWindowOptions` allows customization of the created window.
   By default, child windows are closed when their opener is closed. This can be
-  changed by specifying `outlivesOpener: true`, in which case the opened window
-  will not be closed when its opener is closed.
+  changed by specifying `outlivesOpener: true`, in which case the opened window will not be closed when its opener is closed.
+  Specifying `createWindow` allows to customize the way that window is created.
   Returning an unrecognized value such as a null, undefined, or an object
   without a recognized 'action' value will result in a console error and have
   the same effect as returning `{action: 'deny'}`.
```
This paragraph is more or less repeating the docs in WindowOpenHandlerResponse, we should probably remove this and fold it into window-open-handler-response.md.

nornagon reviewed (approved):
I think this is good to go. Just one minor docs comment.

khalwa commented:
Would it be possible to merge this as it currently is and I'd do the documentation change as a separate PR? Unfortunately the organisation marked the fork repository as read only. If that is not acceptable them I'll try to set up another branch from my own electron repo fork with this commit cherypicked and create a separate PR for this change. Please let me know what you think about this.

samuelmaddock commented:
> Would it be possible to merge this as it currently is and I'd do the documentation change as a separate PR? Unfortunately the organisation marked the fork repository as read only. If that is not acceptable then I'll try to set up another branch from my own electron repo fork with this commit cherypicked and create a separate PR for this change. Please let me know what you think about this.

@khalwa It would be preferred to have this ready prior to merging. Happy to help land this if needed!

khalwa commented:
@samuelmaddock @nornagon I have set up a copy of this PR here with updated documentation: https://github.com/electron/electron/pull/41432 Please let me know if change I did is okay for you on that PR. I'm closing this one.
